presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,268 @@
+                                .collect(groupingBy(Map.Entry::getKey, mapping(Map.Entry::getValue, toImmutableList())));
+                        for (Map.Entry<String, ImmutableList<TupleDomain<String>>> entry : stageDynamicTupleDomains.entrySet()) {
+                            List<TupleDomain<String>> sourceDynamicTupleDomains = entry.getValue();
+                            // check if all tasks of a dynamic filter source have reported TupleDomain summary
+                            if (tasks.size() == sourceDynamicTupleDomains.size()) {
+                                dynamicFilterSummaries.computeIfAbsent(
+                                        SourceDescriptor.of(queryId, entry.getKey()),
+                                        ignored -> SettableFuture.create())
+                                        .set(columnWiseUnion(sourceDynamicTupleDomains));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public ListenableFuture<TupleDomain<String>> getSummary(QueryId queryId, String source)
+    {
+        return dynamicFilterSummaries.computeIfAbsent(SourceDescriptor.of(queryId, source), ignored -> SettableFuture.create());
+    }
+
+    public void removeQuery(QueryId queryId)
+    {
+        dynamicFilterSummaries.keySet().removeIf(sourceDescriptor -> sourceDescriptor.getQueryId().equals(queryId));
+        queries.remove(queryId);
+    }
+
+    public Supplier<TupleDomain<ColumnHandle>> createDynamicFilterSupplier(QueryId queryId, List<DynamicFilters.Descriptor> dynamicFilters, Map<Symbol, ColumnHandle> columnHandles, Map<String, PlanNodeId> dynamicFilterSources)
+    {
+        AtomicReference<TupleDomain<ColumnHandle>> dynamicPredicates = new AtomicReference<>(TupleDomain.all());
+        dynamicFilters.stream()
+                .map(DynamicFilters.Descriptor::getId)
+                .forEach(filterId -> {
+                    ListenableFuture<TupleDomain<ColumnHandle>> dynamicFilterFuture = Futures.transform(
+                            getSummary(queryId, dynamicFilterSources.get(filterId).toString()),
+                            summary -> translateSummaryToTupleDomain(summary, dynamicFilters, columnHandles),
+                            directExecutor());
+                    addSuccessCallback(
+                            dynamicFilterFuture, tupleDomain -> dynamicPredicates.updateAndGet(
+                                    // Intersect dynamic filters' predicates when they become ready
+                                    constraint -> constraint.intersect(tupleDomain)));
+                });
+        return dynamicPredicates::get;
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        collectDynamicFiltersExecutor.shutdownNow();
+    }
+
+    private static List<StageInfo> getStagesWithDynamicFilters(List<StageInfo> stages)
```

Comment:
I think you can remove this method and enumerate all stages instead

Reply:
Removed